src/selectors/layers.js

```diff
@@ -0,0 +1,51 @@
+import { createSelector } from 'reselect';
+import { getLayoutNodes, getGraphSize } from './layout';
+import { getVisibleLayerIDs } from './disabled';
+
+const getLayerName = state => state.layer.name;
+
+/**
+ * Get layer positions
+ */
+export const getLayers = createSelector(
+  [getLayoutNodes, getVisibleLayerIDs, getLayerName, getGraphSize],
+  (nodes, layerIDs, layerName, { width }) => {
+    // Get list of layer Y positions from nodes
+    const layerY = nodes.reduce((layerY, node) => {
+      if (!layerY[node.layer]) {
+        layerY[node.layer] = [];
+      }
+      layerY[node.layer].push(node.y);
+      return layerY;
+    }, {});
+
+    /**
+     * Determine the y position and height of a layer band
+     * @param {number} id
+     */
+    const calculateYPos = (layerID, prevID, nextID) => {
+      const yMin = Math.min(...layerY[layerID]);
```

Comment:
Yeah I'd like to precalculate it. I'd like to precalculate layout too. This needs to be recalculated every time layout changes, so it makes sense to do them together, especially because layout takes approx 50-500ms, while this calc only takes ~1ms.

However that's a lot of possible combinations to precalculate. What is that, 2^n?

Interim idea: Maybe I should store layer calcs and previously-calculated layouts in a store with a unique key so that I can reuse them later if they come up again? Could even possibly do some of the more common variations.

Reply:
I see what you mean. Actually it was more that it looks as though it would run calculations in here twice for each layer (since inputs are previous / next), but thinking about it now, I saw there's only a small number of layers in practice, so it's not going to be an issue.

Reply:
ohhhhhhh you're just talking about the `yMin`/`yMax` calculations! That makes sense. Yeah I could do that actually. idk whether it's would save enough time to be worth optimising though - we're talking about fractions of milliseconds here in total.